### PR TITLE
Optimized memory usage

### PR DESCRIPTION
**Pull request by lubieplacki**
Fixed memory usage so now setup for sha examples finishes correctly.

### CODE REVIEW

File: lib/wraplibsnark.cpp [hunk 1]
```diff
@@ -83,35 +83,42 @@ r1cs_ppzksnark_constraint_system<libff::alt_bn128_pp> createConstraintSystem(con
...
-      libff::bigint<libff::alt_bn128_r_limbs> value = libsnarkBigintFromBytes(A+row*variables*32 + idx*32);
+    while (A_id < A_len && A_vvmap[A_id].constraint_id == row) {
+      libff::bigint<libff::alt_bn128_r_limbs> value = libsnarkBigintFromBytes(A_vvmap[A_id].variable_value);
       libff::alt_bn128_pp::init_public_params();
-      // cout << "C entry " << idx << " in row " << row << ": " << value << endl;
-      if (!value.is_zero()) {
-        // cout << "A(" << idx << ", " << value << ")" << endl;
-        lin_comb_A.add_term(idx,value);
-      }
+      if (!value.is_zero())
```
Comment (Schaeff) [inline lib/wraplibsnark.cpp]: @lubieplacki could this be an assert as we're not supposed to be passing non zero values?

File: lib/wraplibsnark.cpp [hunk 2]
```diff
@@ -83,35 +83,42 @@ r1cs_ppzksnark_constraint_system<libff::alt_bn128_pp> createConstraintSystem(con
...
-    for (int idx=0; idx<variables; idx++) {
-      libff::bigint<libff::alt_bn128_r_limbs> value = libsnarkBigintFromBytes(B+row*variables*32 + idx*32);
-      // cout << "B entry " << idx << " in row " << row << ": " << value << endl;
-      if (!value.is_zero()) {
-        // cout << "B(" << idx << ", " << value << ")" << endl;
-        lin_comb_B.add_term(idx, value);
-      }
+    while (B_id < B_len && B_vvmap[B_id].constraint_id == row) {
+      libff::bigint<libff::alt_bn128_r_limbs> value = libsnarkBigintFromBytes(B_vvmap[B_id].variable_value);
+      libff::alt_bn128_pp::init_public_params();
```
Comment (Schaeff) [inline lib/wraplibsnark.cpp]: @lubieplacki any reason why this init needs to be done again?

File: src/libsnark.rs [hunk 3]
```diff
@@ -51,21 +54,68 @@ pub fn setup<T: Field> (
...
+      for &(idx, ref val) in &b[row] {
+          b_vec.push((row as i32, idx as i32, vec_as_u8_32_array(&val.into_byte_vector())));
+      }
+      for &(idx, ref val) in &c[row] {
+          c_vec.push((row as i32, idx as i32, vec_as_u8_32_array(&val.into_byte_vector())));
+      }
+    }
+
+    // Convert above A,B,C vectors to byte arrays for cpp
+    let mut a_arr: Vec<u8> = vec![0u8; 40 * a_vec.len()];
```
Comment (Schaeff) [inline src/libsnark.rs]: @lubieplacki could you add a comment or define variables for 40, 4, 8?

File: lib/wraplibsnark.cpp (hunk 1 above)
Comment (Schaeff) [reply lib/wraplibsnark.cpp]: Actually no, non zero is not enforced upstream so let's leave that check on.

File: lib/wraplibsnark.cpp (hunk 2 above)
Comment (Schaeff) [reply lib/wraplibsnark.cpp]: I think calling it once at the beginning of the function body is enough.

File: src/libsnark.rs (hunk 3 above)
Comment (lubieplacki) [reply src/libsnark.rs]: :+1:

File: lib/wraplibsnark.cpp (hunk 2 above)
Comment (lubieplacki) [reply lib/wraplibsnark.cpp]: 👍 Will do, it was called in a loop before and it got me confused.

File: lib/wraplibsnark.cpp (hunk 1 above)
Comment (lubieplacki) [reply lib/wraplibsnark.cpp]: :+1:

**Comment by Schaeff** [comment]
Thanks @lubieplacki !

**Pull request merged**